### PR TITLE
frmw434 -rxcharactercount

### PR DESCRIPTION
**Pull request by cathysiller**
JIRA: https://jira.rax.io/browse/FRMW-434 & https://jira.rax.io/browse/FRMW-1377

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] QE LGTM

# Character Count Highlight
### Before
character-highlight-before

### After
(removed)

# Character Count Directive Fix
### Before
character-count-before

### After
character-count-after

### CODE REVIEW

File: utils/rx-page-objects/src/rxCharacterCount.exercise.js [hunk 1]
```diff
@@ -175,38 +174,5 @@ exports.rxCharacterCount = function (options) {
...
-                });
-
-                it('should clear the over-limit text highlighting when the text is reduced', function () {
-                    component.comment = 'h';
-                    expect(component.overLimitText).to.eventually.equal('');
-                });
-            });
-        }
-
-        after(function () {
```
Comment (CITguy) [inline utils/rx-page-objects/src/rxCharacterCount.exercise.js]: We probably should keep this `after()` function to make sure that the exercises clean up after themselves.

Comment (cathysiller) [inline utils/rx-page-objects/src/rxCharacterCount.exercise.js]: reverted -- logic added back in

File: src/elements/Forms/scripts/rxCharacterCount.js [hunk 2]
```diff
@@ -54,19 +54,14 @@ angular.module('encore.ui.elements')
  * correctly for your uses.
  *
  * ### Highlighting ###
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.js]: Should probably remove this entire section, considering the feature is no longer available.

File: src/elements/Forms/scripts/rxCharacterCount.js [hunk 3]
```diff
@@ -54,19 +54,14 @@ angular.module('encore.ui.elements')
...
  *
  * Known failure cases:
  * * Content that causes a scrollbar in the textarea
  * * Initial text (coming from the model) that is over the limit
  *
  * @param {Number=} [low-boundary=10] How far from the maximum to enter a warning state
- * @param {Number=} [max-characters=254] The maximum number of characters allowed
- * @param {Boolean=} [highlight=false] Whether or not characters over the limit are highlighted
- *
+ * @param {Number=} [max-characters=254] The maximum number of characters allowed *
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.js]: Remove trailing `*`

File: src/elements/Forms/scripts/rxCharacterCount.js [hunk 4]
```diff
@@ -126,49 +113,22 @@ angular.module('encore.ui.elements')
...
-                function () { return ngModelCtrl.$modelValue; },
+                function () { return element[0].value; },
                 function (newValue) {
                     if (typeof newValue !== 'string') {
                         return;
                     }
-                    scope.remaining = maxCharacters - newValue.length;
-                    scope.nearLimit = scope.remaining >= 0 && scope.remaining < lowBoundary;
-                    scope.overLimit = scope.remaining < 0;
+                    scope.$evalAsync(function () {
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.js]: Add comment around why `$evalAsync` was used versus other options.

File: src/elements/Forms/scripts/rxCharacterCount.js [hunk 5]
```diff
@@ -126,49 +113,22 @@ angular.module('encore.ui.elements')
...
+                        if (!attrs.ngTrim || attrs.ngTrim !== 'false') {
+                            newValue = newValue.trim();
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.js]: NOTE TO SELF: double check `.trim()` is supported

File: src/elements/Forms/scripts/rxCharacterCount.spec.js [hunk 6]
```diff
@@ -171,81 +171,4 @@ describe('elements:CharacterCount', function () {
             expect(characterCount.hasClass('ng-hide'), 'character countdown should be visible').to.be.false;
         });
     });
-
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.spec.js]: Given that we moved logic around turning off `ngTrim`, do we need to add additional unit tests to ensure it behaves correctly?

**Comment by jj4th** [comment]
QE-LGTM

File: src/elements/Forms/scripts/rxCharacterCount.js (hunk 2 above)
Comment (cathysiller) [inline src/elements/Forms/scripts/rxCharacterCount.js]: removed along with lines 59-61 in relation to

File: src/elements/Forms/scripts/rxCharacterCount.js (hunk 3 above)
Comment (cathysiller) [inline src/elements/Forms/scripts/rxCharacterCount.js]: removed

File: src/elements/Forms/scripts/rxCharacterCount.spec.js [hunk 7]
```diff
@@ -1,10 +1,14 @@
...
     var originalScope, scope, compile, el, $timeout, $rootScope;
     var initialTemplate = '<textarea ng-model="initComment" rx-character-count></textarea>';
     var defaultTemplate = '<textarea ng-model="comment" rx-character-count></textarea>';
     var maxCharsTemplate = '<textarea ng-model="comment" rx-character-count max-characters="50"></textarea>';
     var boundaryTemplate = '<textarea ng-model="comment" rx-character-count max-characters="20" low-boundary="5">' +
                            '</textarea>';
+    var ngTrimTemplate = '<textarea ng-model="trimComment" rx-character-count max-characters="50" ' +
+                         'ng-trim="true"' + trim + '"></textarea>';
+    var ngTrimFalseTemplate = '<textarea ng-model="trimComment" rx-character-count max-characters="50" ' +
+                              'ng-trim="false"' + trim + '"></textarea>';
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.spec.js]: Please remove the `trim` variable.

File: src/elements/Forms/scripts/rxCharacterCount.spec.js [hunk 8]
```diff
@@ -1,10 +1,14 @@
-describe('elements:CharacterCount', function () {
+describe('elements:CharacterCount', function (trim) {
     var originalScope, scope, compile, el, $timeout, $rootScope;
     var initialTemplate = '<textarea ng-model="initComment" rx-character-count></textarea>';
     var defaultTemplate = '<textarea ng-model="comment" rx-character-count></textarea>';
     var maxCharsTemplate = '<textarea ng-model="comment" rx-character-count max-characters="50"></textarea>';
     var boundaryTemplate = '<textarea ng-model="comment" rx-character-count max-characters="20" low-boundary="5">' +
                            '</textarea>';
+    var ngTrimTemplate = '<textarea ng-model="trimComment" rx-character-count max-characters="50" ' +
+                         'ng-trim="true"' + trim + '"></textarea>';
```
Comment (CITguy) [inline src/elements/Forms/scripts/rxCharacterCount.spec.js]: Please remove the `trim` variable.

Comment (cathysiller) [inline src/elements/Forms/scripts/rxCharacterCount.spec.js]: removed

File: src/elements/Forms/scripts/rxCharacterCount.spec.js (hunk 7 above)
Comment (cathysiller) [inline src/elements/Forms/scripts/rxCharacterCount.spec.js]: removed

**Pull request merged**